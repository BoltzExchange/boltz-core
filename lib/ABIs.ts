import ERC20ABI from '../out/ERC20.sol/ERC20.json';
import ERC20SwapABI from '../out/ERC20Swap.sol/ERC20Swap.json';
import EtherSwapABI from '../out/EtherSwap.sol/EtherSwap.json';
import RouterABI from '../out/Router.sol/Router.json';

export const ContractABIs = {
  ERC20: ERC20ABI.abi,
  ERC20Swap: ERC20SwapABI.abi,
  EtherSwap: EtherSwapABI.abi,
  Router: RouterABI.abi,
} as const;
