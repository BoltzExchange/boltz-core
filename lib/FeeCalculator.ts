import Errors from './consts/Errors';
import { OutputType } from './consts/Enums';

export type Input = {
  type: OutputType,

  // In case the input is a Swap this fields have to be set
  swapDetails?: {
    redeemScript: Buffer;
    preimage?: Buffer;
  };
};

export type Output = {
  type: OutputType;

  isSh?: boolean;
};

// Estimations for the vbytes of different PKH inputs
const inputVbytesEstimations = {
  [OutputType.Bech32]: 108 + (41 * 4),
  [OutputType.Compatibility]: 108 + (64 * 4),
  [OutputType.Legacy]: 148 * 4,
};

// Estimations for the vbytes of different PKH and SH outputs
const outputVbytesEstimations = {
  pkh: {
    [OutputType.Bech32]: 31 * 4,
    [OutputType.Legacy]: 34 * 4,
    [OutputType.Taproot]: 43 * 4,
  },
  sh: {
    [OutputType.Bech32]: 44 * 4,
    [OutputType.Legacy]: 32 * 4,
  },
};

const estimateInput = (input: Input) => {
  if (input.type === OutputType.Taproot) {
    throw Errors.TAPROOT_NOT_SUPPORTED;
  }

  if (input.swapDetails) {
    const swapSize = [
      // PUSHDATA opcode
      1,
      // ECDSA signature
      72,
      // PUSHDATA opcode if there is a preimage
      input.swapDetails.preimage ? 1 : 0,
      // preimage if there is one
      input.swapDetails.preimage ? input.swapDetails.preimage.length : 0,
      // Sequence
      4,
      // Redeemscript
      input.swapDetails.redeemScript.length,
    ].reduce((swapSize, n) => swapSize + n);

    switch (input.type) {
      case OutputType.Bech32:
        return 41 * 4 + swapSize;

      case OutputType.Compatibility:
        return ((41 + 35) * 4) + swapSize;

      case OutputType.Legacy:
        return (swapSize + 41) * 4;
    }

    return 0;
  } else {
    return inputVbytesEstimations[input.type];
  }
};

const estimateOutput = (output: Output): number => {
  if (output.type === OutputType.Compatibility) {
    return outputVbytesEstimations.sh[OutputType.Legacy];
  }

  if (output.isSh) {
    return outputVbytesEstimations.sh[output.type];
  } else {
    return outputVbytesEstimations.pkh[output.type];
  }
};

/**
 * Estimates the vbytes of a transaction
 */
export const estimateSize = (inputs: Input[], outputs: Output[]): number => {
  // A raw transaction has always 4 bytes for the version and 4 for the locktime
  let sum = 8 * 4;
  let hasWitness = false;

  inputs.forEach((input) => {
    if (input.type !== OutputType.Legacy) {
      hasWitness = true;
    }

    sum += estimateInput(input);
  });

  outputs.forEach((output) => {
    sum += estimateOutput(output);
  });

  if (hasWitness) {
    sum += 8;
  }

  // Divide the weight by 4 and round it up to the next integer
  return Math.ceil(sum / 4);
};

/**
 * Estimates the amount of satoshis that should be paid as fee
 */
export const estimateFee = (satsPerVbyte: number, inputs: Input[], outputs: Output[]): number => {
  const size = estimateSize(inputs, outputs);

  return size * satsPerVbyte;
};
