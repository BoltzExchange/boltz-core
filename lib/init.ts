import { initEccLib } from 'bitcoinjs-lib';
import { TinySecp256k1Interface } from 'bitcoinjs-lib/src/types';

export const init = (eccLib: TinySecp256k1Interface) => initEccLib(eccLib);
