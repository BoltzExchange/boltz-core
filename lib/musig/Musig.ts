import { ECPairInterface } from 'ecpair';
import { toXOnly } from 'bitcoinjs-lib/src/psbt/bip371';
import { Secp256k1ZKP } from '@michael1011/secp256k1-zkp';

class Musig {
  private readonly pubkeyAgg: {
    aggPubkey: Uint8Array;
    keyaggCache: Uint8Array;
  };
  private readonly nonce: {
    pubNonce: Uint8Array;
    secNonce: Uint8Array;
  };

  private readonly myIndex: number;
  private readonly partialSignatures: (Uint8Array | null)[];

  private pubNonces?: Uint8Array[];

  private nonceAgg?: Uint8Array;
  private session?: Uint8Array;

  constructor(
    private readonly secp: Secp256k1ZKP,
    private readonly key: ECPairInterface,
    public readonly sessionId: Buffer,
    public readonly publicKeys: Buffer[],
  ) {
    if (publicKeys.length < 2) {
      throw 'need at least 2 keys to aggregate';
    }

    this.myIndex = this.publicKeys.findIndex((key) =>
      this.key.publicKey.equals(key),
    );

    if (this.myIndex === -1) {
      throw 'our key is not publicKeys';
    }

    this.pubkeyAgg = this.secp.musig.pubkeyAgg(publicKeys.map(toXOnly));
    this.nonce = this.secp.musig.nonceGen(sessionId);
    this.partialSignatures = Array(publicKeys.length).fill(null);
  }

  public numParticipants = (): number => {
    return this.publicKeys.length;
  };

  public getAggregatedPublicKey = (): Buffer => {
    return Buffer.from(this.pubkeyAgg.aggPubkey);
  };

  public getPublicNonce = (): Uint8Array => {
    return this.nonce.pubNonce;
  };

  public tweakKey = (tweak: Uint8Array): Buffer => {
    const tweaked = this.secp.musig.pubkeyXonlyTweakAdd(
      this.pubkeyAgg.keyaggCache,
      tweak,
      true,
    );
    this.pubkeyAgg.keyaggCache = tweaked.keyaggCache;
    return Buffer.from(tweaked.pubkey);
  };

  public aggregateNoncesOrdered = (nonces: Uint8Array[]) => {
    if (this.publicKeys.length !== nonces.length) {
      throw 'number of nonces != number of public keys';
    }

    const myNonceIndex = nonces.findIndex((nonce) =>
      Buffer.from(this.nonce.pubNonce).equals(nonce),
    );
    if (myNonceIndex !== this.myIndex) {
      throw 'our nonce is at incorrect index';
    }

    this.pubNonces = nonces;
    this.nonceAgg = this.secp.musig.nonceAgg(nonces);
  };

  public aggregateNonces = (nonces: Map<Uint8Array, Uint8Array>) => {
    const ordered: Uint8Array[] = [];

    if (!nonces.has(this.key.publicKey)) {
      nonces.set(this.key.publicKey, this.getPublicNonce());
    }

    if (this.publicKeys.length !== nonces.size) {
      throw 'number of nonces != number of public keys';
    }

    for (const key of this.publicKeys) {
      const nonce = nonces.get(key);
      if (nonce === undefined) {
        throw `could not find nonce for public key ${key.toString('hex')}`;
      }

      ordered.push(nonce);
    }

    this.aggregateNoncesOrdered(ordered);
  };

  public initializeSession = (msg: Uint8Array) => {
    if (this.nonceAgg === undefined) {
      throw 'nonces not aggregated';
    }

    if (this.session !== undefined) {
      throw 'session already initialized';
    }

    this.session = this.secp.musig.nonceProcess(
      this.nonceAgg,
      msg,
      this.pubkeyAgg.keyaggCache,
    );
  };

  /**
   * Returns our partial signature and adds it to the internal list
   */
  public signPartial = () => {
    if (this.session === undefined) {
      throw 'session not initialized';
    }

    const sig = this.secp.musig.partialSign(
      this.nonce.secNonce,
      this.key.privateKey!,
      this.pubkeyAgg.keyaggCache,
      this.session,
    );
    this.partialSignatures[this.myIndex] = sig;

    return sig;
  };

  public verifyPartial = (
    publicKeyOrIndex: Buffer | number,
    signature: Uint8Array,
  ): boolean => {
    if (this.pubNonces === undefined) {
      throw 'public nonces missing';
    }

    if (this.session === undefined) {
      throw 'session not initialized';
    }

    const publicKey =
      typeof publicKeyOrIndex === 'number'
        ? this.publicKeys[publicKeyOrIndex]
        : publicKeyOrIndex;
    const index = this.indexOfPublicKeyOrIndex(publicKey);

    return this.secp.musig.partialVerify(
      signature,
      this.pubNonces[index],
      toXOnly(publicKey),
      this.pubkeyAgg.keyaggCache,
      this.session,
    );
  };

  /**
   * Adds a partial signature after verifying it
   */
  public addPartial = (
    publicKeyOrIndex: Buffer | number,
    signature: Uint8Array,
  ) => {
    if (!this.verifyPartial(publicKeyOrIndex, signature)) {
      throw 'invalid partial signature';
    }

    this.partialSignatures[this.indexOfPublicKeyOrIndex(publicKeyOrIndex)] =
      signature;
  };

  public aggregatePartials = (): Buffer => {
    if (this.session === undefined) {
      throw 'session not initialized';
    }

    if (this.partialSignatures.some((partial) => partial === null)) {
      throw 'not all partial signatures are set';
    }

    return Buffer.from(
      this.secp.musig.partialSigAgg(
        this.session,
        this.partialSignatures as Uint8Array[],
      ),
    );
  };

  private indexOfPublicKeyOrIndex = (
    publicKeyOrIndex: Buffer | number,
  ): number => {
    const index =
      typeof publicKeyOrIndex === 'number'
        ? publicKeyOrIndex
        : this.publicKeys.findIndex((key) => publicKeyOrIndex.equals(key));

    if (index === -1) {
      throw `could not find index of public key ${Buffer.from(
        publicKeyOrIndex as Uint8Array,
      ).toString('hex')}`;
    }

    if (index > this.publicKeys.length - 1) {
      throw 'index out of range';
    }

    return index;
  };
}

export default Musig;
