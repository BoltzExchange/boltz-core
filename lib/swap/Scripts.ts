/**
 * This file is based on the repository github.com/submarineswaps/swaps-service created by Alex Bosworth
 */

import ops from '@boltz/bitcoin-ops';
import { script, crypto } from 'bitcoinjs-lib';

/**
 * Get a P2WPKH output script
 *
 * @param hash public key hash hex Buffer
 *
 * @returns P2WPKH output script Buffer
 */
export const p2wpkhOutput = (hash: Buffer) => {
  return script.compile([
    ops.OP_0,
    hash,
  ]);
};

/**
 * Encode a P2WSH output script
 *
 * @param scriptHex redeem script hex Buffer
 *
 * @returns P2WSH output script Buffer
 */
export const p2wshOutput = (scriptHex: Buffer) => {
  return script.compile([
    ops.OP_0,
    crypto.sha256(scriptHex),
  ]);
};

/**
 * Get a P2PKH output script
 *
 * @param hash public key hash hex Buffer
 *
 * @returns P2PKH output script Buffer
 */
export const p2pkhOutput = (hash: Buffer) => {
  return script.compile([
    ops.OP_DUP,
    ops.OP_HASH160,
    hash,
    ops.OP_EQUALVERIFY,
    ops.OP_CHECKSIG,
  ]);
};

/**
 * Encode a P2SH output script
 *
 * @param scriptHex redeem script hex Buffer
 *
 * @returns P2SH output script Buffer
 */
export const p2shOutput = (scriptHex: Buffer) => {
  return script.compile([
    ops.OP_HASH160,
    crypto.hash160(scriptHex),
    ops.OP_EQUAL,
  ]);
};

/**
 * Get a P2SH nested P2WPKH output script
 *
 * @param hash public key hash hex Buffer
 */
export const p2shP2wpkhOutput = (hash: Buffer) => {
  const witness = p2wpkhOutput(hash);

  return {
    redeemScript: witness,
    outputScript: p2shOutput(witness),
  };
};

/**
 * Get a P2SH nested P2WSH output script
 *
 * @param scriptHex redeem script hex Buffer
 */
export const p2shP2wshOutput = (scriptHex: Buffer) => {
  const witness = p2wshOutput(scriptHex);

  return p2shOutput(witness);
};
