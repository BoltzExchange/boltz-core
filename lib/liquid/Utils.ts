import { confidential, TxOutput } from 'liquidjs-lib';
import { confidentialLiquid } from './init';

export const getOutputValue = (
  output: TxOutput & {
    blindingPrivateKey?: Buffer;
  },
): number => {
  return output.blindingPrivateKey
    ? Number(
        confidentialLiquid.unblindOutputWithKey(
          output,
          output.blindingPrivateKey,
        ).value,
      )
    : confidential.confidentialValueToSatoshi(output.value);
};
