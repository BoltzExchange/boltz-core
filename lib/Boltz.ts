import Networks from './consts/Networks';
import * as Scripts from './swap/Scripts';
import { OutputType } from './consts/Enums';
import * as SwapUtils from './swap/SwapUtils';
import { swapScript } from './swap/SwapScript';
import { detectSwap } from './swap/SwapDetector';
import { constructClaimTransaction } from './swap/Claim';
import { detectPreimage } from './swap/PreimageDetector';
import { constructRefundTransaction } from './swap/Refund';
import { estimateFee, estimateSize } from './FeeCalculator';
import { ScriptElement, TransactionOutput } from './consts/Types';

export {
  Networks,
  OutputType,
  ScriptElement,
  TransactionOutput,

  Scripts,

  swapScript,

  detectSwap,
  detectPreimage,

  estimateFee,
  estimateSize,

  constructClaimTransaction,
  constructRefundTransaction,

  SwapUtils,
};
