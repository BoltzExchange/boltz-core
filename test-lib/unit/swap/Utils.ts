import Errors from '../../../lib/consts/Errors';
import { OutputType } from '../../../lib/consts/Enums';
import { p2wshOutput, p2shOutput, p2shP2wshOutput } from '../../../lib/swap/Scripts';

export const getScriptHashFunction = (type: OutputType): (scriptHex: Buffer) => Buffer => {
  switch (type) {
    case OutputType.Bech32: return p2wshOutput;
    case OutputType.Legacy: return p2shOutput;
    case OutputType.Compatibility: return p2shP2wshOutput;
    case OutputType.Taproot: throw Errors.TAPROOT_NOT_SUPPORTED;
  }
};
