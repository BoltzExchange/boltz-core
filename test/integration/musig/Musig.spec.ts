import { randomBytes } from 'crypto';
import * as ecc from 'tiny-secp256k1';
import { address, initEccLib, Transaction } from 'bitcoinjs-lib';
import zkp, { Secp256k1ZKP } from '@vulpemventures/secp256k1-zkp';
import { bitcoinClient } from '../Utils';
import { ECPair } from '../../unit/Utils';
import Musig from '../../../lib/musig/Musig';
import { Networks } from '../../../lib/Boltz';
import { getHexBuffer } from '../../../lib/Utils';
import { p2trOutput } from '../../../lib/swap/Scripts';

describe('Musig', () => {
  let secp: Secp256k1ZKP;

  beforeAll(async () => {
    secp = await zkp();
    initEccLib(ecc);
    await bitcoinClient.init();
  });

  afterAll(async () => {
    await bitcoinClient.generate(1);
  });

  test('should create P2TR addresses', () => {
    const ourKey = ECPair.fromPrivateKey(
      getHexBuffer(
        '1f1d1a9d8c3507b251ff29096511ffe66a75437e3bcdf912e7433e55e87f569c',
      ),
    );
    const musig = new Musig(secp, ourKey, randomBytes(32), [
      ourKey.publicKey,
      getHexBuffer(
        '02e88bd3780532bbb4a127a2e041467bc42c5cc4ff16ddb5afa7e27c5b653de44c',
      ),
    ]);
    expect(
      address.fromOutputScript(
        p2trOutput(musig.getAggregatedPublicKey()),
        Networks.bitcoinRegtest,
      ),
    ).toMatchSnapshot();
  });

  test('should spend Musig P2TR outputs', async () => {
    const ourKey = ECPair.makeRandom();
    const theirKey = ECPair.makeRandom();

    const musig = new Musig(secp, ourKey, randomBytes(32), [
      ourKey.publicKey,
      theirKey.publicKey,
    ]);

    // Fund address
    const outputScript = p2trOutput(musig.getAggregatedPublicKey());
    const amount = 10_000;

    const txId = await bitcoinClient.sendToAddress(
      address.fromOutputScript(outputScript, Networks.bitcoinRegtest),
      amount,
    );
    const prevTx = Transaction.fromHex(
      await bitcoinClient.getRawTransaction(txId),
    );
    const outputIndex = prevTx.outs.findIndex((out) =>
      out.script.equals(outputScript),
    );

    // Construct transaction
    const tx = new Transaction();
    tx.addInput(prevTx.getHash(), outputIndex);
    tx.addOutput(
      address.toOutputScript(
        await bitcoinClient.getNewAddress(),
        Networks.bitcoinRegtest,
      ),
      amount - 1_000,
    );

    const sigHash = tx.hashForWitnessV1(
      0,
      [outputScript],
      [amount],
      Transaction.SIGHASH_DEFAULT,
    );

    // Create signature
    const theirNonce = secp.musig.nonceGen(randomBytes(32), theirKey.publicKey);
    musig.aggregateNonces([[theirKey.publicKey, theirNonce.pubNonce]]);
    musig.initializeSession(sigHash);
    musig.signPartial();
    musig.addPartial(
      theirKey.publicKey,
      secp.musig.partialSign(
        theirNonce.secNonce,
        theirKey.privateKey!,
        musig['pubkeyAgg'].keyaggCache,
        musig['session']!,
      ),
    );

    // Finalize and broadcast
    tx.setWitness(0, [musig.aggregatePartials()]);

    await bitcoinClient.sendRawTransaction(tx.toHex());
  });
});
