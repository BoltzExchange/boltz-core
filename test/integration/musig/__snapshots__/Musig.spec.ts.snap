// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Musig should create P2TR addresses 1`] = `"bcrt1pn8rm287f57grctmr3xwt8lrsu2xf4pg5a0zaexjznp89v6me88pq5rut2m"`;
